テーブル設計

shivonnu opened the pull request:
# What
テーブル設計
#  Why
これから作るアプリのテーブル設計のため
以下ER図です。
![furima_app_er (6)](https://user-images.githubusercontent.com/69893922/93011773-3d632100-f5d4-11ea-884c-ab78d30b1494.png)

shivonnu committed:
imageDL追加、gemrubocop導入

shivonnu committed:
スタイルシート、ビュー素材貼り付け

shivonnu committed:
テーブル設計

Katsuno0814 reviewed (commented):
こちらのプルリクエストですが、プルリクエストにWhatとWhyが書かれていません。
What（どのような実装をしているのか）と Why（なぜこの実装が必要なのか）を書くことで、
このブランチは何を実装していて、なぜその実装を行ったのかが他の人に伝えることができます。
What をしっかり書くことで、他者にコードを見てもらう際に、スムーズな理解を促すことができ、Whyを書くことでなぜこの実装を行ったのか、本来の目的に沿ったレビュアーとのコミュニケーションが可能となります。
お手数かけますが、記入後再度提出いただけますよう、よろしくお願いします。
https://master.tech-camp.in/curriculums/4675

shivonnu commented:
ご指摘ありがとうございます。
修正しましたので再度コードレビュー依頼させて頂きます。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| last_name_kana  | string | null: false |
+| fast_name_kana  | string | null: false |
+| name_katakana   | string | null: false |
+| birthday_year   | date   | null: false |
+| birthday_month  | date   | null: false |
+| birthday_day    | date   | null: false |
```
こちら誕生日のカラムですがdate型を使われているので年、月、日を1つで指定して取得することができます。
よってカラムは1つにまとめましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| product_status       | integer | null: false                    |
+| price                | integer | null: false                    |
+| users_id             | integer | null: false, foreign_key: true |
 
```
商品のテーブルにて、出品時に必要なカラムがまだ足りないようです。
見本アプリを参考に保存するカラムを設計しましょう！

tenpadrummer replied:
配送料の負担, 発送元の地域,  発送までの日数のカラムも追加しましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
+| product_name         | string  | null: false                    |
+| exhibition_image     | string  | null: false                    |
```
今回の商品の出品の際の画像投稿には、カリキュラムで学習した「active_storage」を使用します。
その際、テーブルを用意する必要はないので、設計の段階からこのテーブルを省いても問題ありません。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+|----------------------|---------|--------------------------------|
+| product_name         | string  | null: false                    |
+| exhibition_image     | string  | null: false                    |
+| product_description  | integer | null: false                    |
```
商品説明を保存するカラムは、説明が長くなることを仮定すると、text型で設計した方が無難かと考えます。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+### Association
 
-* How to run the test suite
+- belongs_to :users
```
belongs_toの場合単数系で書きましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
+| credit_card_id       | integer | null: false                    |
+| expiation_date_month | integer | null: false                    |
+| expiation_date_year  | integer | null: false                    |
+| security_code        | integer | null: false                    |
```
セキュリティの観点から、クレジットカードの情報はデータベースに直接保存しないようにしましょう。

また、今回のフリマアプリではカードを登録する際は購入の度に入力となり、データベースにわざわざ記録しなくても良い情報となります。
ですので、これらのカラムを用意しなくても大丈夫です！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+### Association
+
+- belongs_to :users
 
```
productに対してのアソシエーションを記述いたしましょう。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| address              | string  | null: false                    |
+| building_name        | string  | null: false                    |
+| phone_number         | integer |                                |
+| users_id             | integer | null: false, foreign_key: true |
+| products_id          | integer | null: false, foreign_key: true |
```
| users_id             | integer | null: false, foreign_key: true |
| products_id          | integer | null: false, foreign_key: true |
こちら住所のカラムは購入管理（ Purchase テーブル）と結びつくのが無難かと思われますので
上記２つはさくじょして Purchase テーブルを参照する様にしましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| city                 | string  | null: false                    |
+| address              | string  | null: false                    |
+| building_name        | string  | null: false                    |
+| phone_number         | integer |                                |
```
電話番号は数字の文字列として実装しておきたいためstring型で実装しましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| prefectures          | string  | null: false                    |
+| city                 | string  | null: false                    |
+| address              | string  | null: false                    |
+| building_name        | string  | null: false                    |
```
今回の実装では、建物番号の入力は任意になるため、not null制約はなくても大丈夫です！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+|----------------------|---------|--------------------------------|
+| postal_code          | integer |                                |
+| expiration_date      | integer | null: false                    |
+| prefectures          | string  | null: false                    |
```
都道府県のカラムはactive_hashで実装の予定のため、integer型にしましょう！
詳しくは下記のカリキュラムを参考にしてください！
https://master.tech-camp.in/curriculums/4753

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
+| postal_code          | integer |                                |
```
郵便番号はハイフンを必要とすることも考慮してstring型で実装しましょう！
またnull: false  は必要になります！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,90 @@
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
+| postal_code          | integer |                                |
+| expiration_date      | integer | null: false                    |
```
| expiration_date      | integer | null: false                    |
こちらに記載するのではなく商品テーブルの方に記載いたしましょう！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

shivonnu committed:
誕生日のカラムを一つにまとめる

誕生日のカラムを年、月、日で分けてていたのを一つにまとめました。

shivonnu committed:
商品のテーブルにカテゴリーを追加

商品のテーブルに出品時に必要なカラムが不足してたので、カテゴリーカラムを追加しました。

shivonnu committed:
画像投稿カラムの削除

商品出品の際 「active_strage」を使う為にテーブルに保存する必要が無い画像投稿カラムを削除しました。

shivonnu committed:
商品説明保存カラムをtext型に変更

商品説明を保存するカラムは説明が長くなる可能性がある為、string型からtext型に変更しました。

shivonnu committed:
belongs_toの場合単数形に修正

belongs_toの部分で複数形を使用していたので単数形に修正しました。

shivonnu committed:
クレジットカード情報の削除

セキュリティの観点からクレジットカードの情報はデータベースに直接保存しない事と、今回のフリマアプリではカードを登録する際は購入の度に入力となるので記録する必要が無い事から、カードに関する情報を削除しました。

shivonnu committed:
product、shipping_addressへのアソシエーション記述

purchaseテーブルのproductに対してのアソシエーションを追加しました。shipping_addressもER図では繋いだがREADMEに無かったので追加しました。

shivonnu committed:
user_idとprodusts_idの削除

住所のカラムは購入管理(Purchase)テーブルと結びつくのが無難と思われるので削除してpurchaseテーブルを参照するようにしました。

shivonnu committed:
電話番号をstring型に変更

電話番号は数字の文字列として実装しておきたい為string型に変更しました。

shivonnu committed:
建物番号のnot null制約を削除

今回の実装では建物番号の入力は任意になる為、not null制約は削除しました。

shivonnu committed:
都道府県をinteger型に変更

都道府県のカラムはactive_hashで実装予定の為integer型に変更しました。

shivonnu committed:
郵便番号をintegerに変更、null:false追加

郵便番号はハイフンを必要とする事を考慮してinteger型→string型に変更しました。null:false必要になる為空欄→追記しました。

shivonnu committed:
有効期限の削除

こちら後から確認した所余計なカラムだった為、削除させて頂きました。

shivonnu committed:
name_katakanaカラム削除

見直した所余計なカラムがあったので削除しました。

shivonnu committed:
カテゴリーカラム位置入れ替え

実際のページの並び順にカラムを入れ替えました。

shivonnu committed:
商品のアソシエーション追加

purchaseからproductsへ片方だけアソシエーション追記したことに気がついたので、productsからpurchaseへも追記しました。

shivonnu commented:
コードレビューありがとうございます。
修正しましたので再度コードレビュー依頼させて頂きます。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
 
-* Configuration
+- has_many   :products
+- belongs_to :purchase
```
1人のユーザーが商品を複数回購入することができると想定すると、購入テーブルとのアソシエーションはhas_manyの使用が好ましいと考えます。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| product_name         | string  | null: false                    |
+| product_description  | text    | null: false                    |
+| category             | string  | null: false                    |
+| product_status       | integer | null: false                    |
```
テーブル名から判断できるので、カラム名のproduct_は不要です！

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+- belongs_to :purchase
 
-* Database creation
+## Products テーブル
```
商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリーのカラムはactive_hashで実装の予定のため、integer型にしましょう！
詳しくは下記のカリキュラムを参考にしてください！（references型は使用しません。）
https://master.tech-camp.in/curriculums/4753

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+- belongs_to :shipping _address
+- belongs_to :purchase
+
+## Delivery_information テーブル
```
active_hashで実装しますので、こちらのテーブルは不要です！

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| city                 | string  | null: false                    |
+| address              | string  | null: false                    |
+| building_name        | string  |                                |
+| phone_number         | string  |                                |
```
電話番号は必須なので、not null 制約をつけておきましょう！

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+- belongs_to :delivery_information
+- belongs_to :purchase
+
+## Purchase テーブル
```
テーブル名は複数形が一般的です。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+- belongs_to :product
+- belongs_to :shipping_address
+
+## Shipping_address テーブル
```
Purchases テーブルの外部キーカラムを用意しましょう！

syuji-yamamoto replied:
テーブル名は複数形が一般的です。
今回の場合は```Shipping_addresses テーブル```としましょう。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,84 @@
+
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
+| users_id             | integer | null: false, foreign_key: true |
+| products_id          | integer | null: false, foreign_key: true |
```
外部キーカラムは単数系が好ましいです。修正しましょう。
他のテーブルの外部キーカラムも複数形になっていたら修正しましょう。
```
user_id 
product_id   
```

syuji-yamamoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点を確認させていただきました！
コメントしましたので、ご確認ください！

shivonnu committed:
利用者と購入テーブルアソシエーションをhas_manyに変更

一人のユーザーが複数回商品を購入できる為、usersからpurchase（購入）テーブル間のアソシエーションをbelongs_to→has manyに変更しました。

shivonnu committed:
カラム名product_の削除

テーブル名から判断できるのでカラムのprodust_部分を削除しました。

shivonnu committed:
カテゴリーのカラムをinteger型に変更

カテゴリーのカラムはactive_hashで実装予定なのでstring型→integer型に変更しました。*商品の状態、配送料の負担、発送元の地域、発送までの日数、カテゴリーはinteger型。references型は使用しない。

shivonnu committed:
Delivery_informationテーブルの削除

active_hashで実装するのでテーブルに保存する必要が無い為削除しました。
同時に繋がっていたアソシエーションも削除しました。

shivonnu committed:
電話番号のnot null制約の追加

電話番号は必須なので空欄→null:falseに変更しました。

shivonnu committed:
Purchaseテーブルを複数形の名前へ変更

テーブル名は複数形が一般的な為Purchase→Purchasesにテーブル名を変更しました。

shivonnu committed:
pruchasesテーブルにshipping_addresses_id追加

購入情報テーブルに配送先情報の外部キーを追加しました。

shivonnu committed:
配送先情報テーブルを複数形に変更

テーブル名は複数形が一般的な為、Shipping_address→Shipping_addressesに変更しました。

shivonnu committed:
外部キーを単数系に修正

外部キーカラムは単数形が好ましい為、productsテーブルとpurchasesテーブル外部キーを修正しました。users_id,products_id→user_id,product_id。

shivonnu commented:
コードレビューありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

miyagit commented:
こちら.DS_Storeは、githubに挙げるのに不要なファイルのため、gitignoreに追加して今後githubに上がらない様にしましょう。
gitignoreの追加の仕方が分からない場合は、「rails gitignore 使い方」でGoogleで検索してみるとやり方が載っているサイトが出てくるので、サイトを参考に実装しましょう！

shivonnu committed:
.DS_Storeが今後githubに上がらないようにgitignoneに追加

.DS_Storeはgithubに不要なファイルの為今後上がらないようにgitignoneに追記しました。

shivonnu commented:
コードレビューありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :user
+- belongs_to :shipping _address
+- belongs_to :purchase
```
```
- belongs_to :shipping _address
- belongs_to :purchase
```

こちら、has_one: purchaseにしておきましょう！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
+
+### Association
+
+- has_many   :products
```
- has_many   :products

は不要です！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
+|-----------------------|---------|--------------------------------|
+| user_id               | integer | null: false, foreign_key: true |
+| product_id            | integer | null: false, foreign_key: true |
+| shipping_address_id   | integer | null: false, foreign_key: true |
```
| shipping_address_id   | integer | null: false, foreign_key: true |
は不要です！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| shipping_address_id   | integer | null: false, foreign_key: true |
+
+### Association
+
```
- belongs_to :user
- has_many   :products
- belongs_to :shipping_address

アソシエーションは以下のようにしましょう！

```
- belongs_to :user
- belongs_to :item
- belongs_to :shipping_address
```

tenpadrummer reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

shivonnu committed:
配送料の負担、発送元の地域、発送までの日数カラムの追加

Productsテーブルに配送料の負担(shipping_area)、発送元の地域(shipping_fee_burden)、発送までの日数カラム(days_to_ship)の追加をしました。

shivonnu committed:
Products,Puechasesテーブル間のアソシエーションをhas_oneに修正

商品は沢山の購入者に買われるので、Products,Puechasesテーブル間のアソシエーションをbelongs_to→has_oneに修正しました。

shivonnu committed:
Products,Shipping_address間のアソシエーションを削除

配送先情報は購入情報と繋がってれば動作できる為Products,Shipping_address間のアソシエーションを削除しました。

shivonnu committed:
shipping_address_id カラムの削除

purchasesテーブルのshipping_address_idカラムを不要な為削除しました。

shivonnu committed:
購入情報テーブルの商品テーブルへのアソシエーションをbelongs_toへ変更

購入情報にとっては対応する商品は一つだけなので、
購入情報テーブルから商品テーブルへのアソシエーションをhas_many→belongs_toへ変更しました。

shivonnu committed:
商品テーブルの外部キーカラムをテーブルの下部に入れ替え

商品productsテーブルの外部キーカラムは下の列にあるほうが望ましいと考えたので場所を入れ替えました。

shivonnu commented:
コードレビューありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| product_id            | integer | null: false, foreign_key: true |
+
+### Association
+
```
大変申し訳ございません。
こちら修正していただきましたが、正しくは下記のアソシエーションになります。

```
belongs_to :user
belongs_to :item
has_one :shipping_address
```

※購入する時のみ配送先を保存する仕様のため、
購入テーブルが親となり、それに紐づく配送先テーブルが子の関係になるためです。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,69 @@
 
-* Services (job queues, cache servers, search engines, etc.)
+- belongs_to :user
+- has_one    :purchases
```
has_oneは単数系にしましょう！

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,69 @@
+## Shipping_addresses テーブル
+
+| column               | Type    | Options                        |
+|----------------------|---------|--------------------------------|
```
Purchasesテーブルの外部キーカラムを用意しましょう！

syuji-yamamoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点を確認させていただきました！
申し訳ございませんが、修正をお願いします！

shivonnu committed:
Purchase,Shipping_address間のテーブルのアソシエーションをhas_oneに変更

購入する時のみ配送先を保存する使用で親と子の関係になるのでPerchase,Shipping_address間のテーブルのアソシエーションをhas_oneに修正。ついでにProductをitemの方がわかりやすそうだったので変更しました。

shivonnu committed:
has_one:purchasesを単数系に修正

has_oneは単数系で扱う必要があるので単数系に修正しました。

shivonnu committed:
Purchasesテーブルとitemsテーブルの外部キーカラムを設定

Purchasesテーブルとitemsテーブルの外部キー制約をreferences型を設定しない為にadd_foreign_keyで設定する書き方に修正。

shivonnu commented:
コードレビューありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| shipping_area        | integer | null: false                                  |
+| shipping_fee_burden  | integer | null: false                                  |
+| days_to_ship         | integer | null: false                                  |
+| user_id              | integer | null: false, add_foreign_key: :items, :users |
```
```
add_foreign_key: :items, :users
```
こちら、foreign_key :trueで大丈夫です！

tenpadrummer commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| column                | Type    | Options                                          |
+|-----------------------|---------|--------------------------------------------------|
+| user_id               | integer | null: false, add_foreign_key: :purchases, :users |
+| item_id               | integer | null: false, add_foreign_key: :purchases, :items |
```
```
add_foreign_key: :purchases, :users
add_foreign_key: :purchases, :items
```

こちら、foreign_key :trueで大丈夫です！

tenpadrummer reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

shivonnu committed:
add_foreign_key:の表記をforeign_key: :trueに戻す修正

不要な修正であった為元に戻しました。add_foreign_key: :items, :users→foreign_key: :true

shivonnu committed:
Purchasesテーブルの外部キーカラムを設定

Purchasesテーブルの外部キーカラムを設定しなくてはならない為、Shipping_addresses テーブルにpurchase_idを設定しました。

shivonnu commented:
コードレビューありがとうございます！
沢山根気よくご指導頂きましてありがとうございます。
修正しましたので再度コードレビュー依頼させて頂きます！

yupimaru commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| column          | Type   | Options     |
+|-----------------|--------|-------------|
+| nickname        | string | null: false |
+| mail_address    | string | null: false |
```
```
| mail_address    | string | null: false |
```

メールアドレスを保存するカラムですが、deviseを用いる場合`email`というカラムが用意されていますので、
こちらのカラム名も`email`としておきましょう。

yupimaru commented on README.md:
```diff
@@ -1,24 +1,70 @@
+|-----------------|--------|-------------|
+| nickname        | string | null: false |
+| mail_address    | string | null: false |
+| password        | string | null: false |
```
```
| password        | string | null: false |
```

パスワードを保存するカラムですが、deviseを用いる場合デフォルトで`encrypted_password`というカラムが用意されていますので、
こちらのカラム名も`encrypted_password`としておきましょう。

yupimaru reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

shivonnu committed:
メールアドレスの名前をmail_adless→emailに変更

deviseを用いる場合はメールアドレスを保存するカラムはemailというカラムが用意されている為、カラム名をemailに修正しました。

shivonnu committed:
passwordカラムをencrypted_passwordに変更

パスワードを保存するカラムには、deviseを用いる場合デフォルトでencrypted_passwordというカラムが用意されているので、password→encrypted_passwordに修正しました。

shivonnu commented:
コードレビューありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

syuji-yamamoto commented:
レビュー依頼ありがとうございます！
修正点を確認させていただきました！

LGTMとさせていただきます！DB設計の実装、本当にお疲れ様でした！
引き続き、実装頑張っていきましょう！

shivonnu commented:
こちらこそ本当にありがとうございます。
頑張ります。
